### PR TITLE
refactor(analytics): Add functional identifyUser API

### PR DESCRIPTION
**Pull request by cshfang**
#### Description of changes

This PR adds the functional identifyUser API to Analytics

#### Description of how you validated changes
Tested against sample app locally

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/analytics/src/types/index.ts [hunk 1]
```diff
@@ -2,5 +2,6 @@
 // SPDX-License-Identifier: Apache-2.0
 
 export * from './Analytics';
+export * from './parameters';
```
Comment (ashwinkumar6) [inline packages/analytics/src/types/index.ts:5]: NIT:
QQ
Should we update these to named exports
We could do this on a separate PR

File: packages/core/src/index.ts [hunk 2]
```diff
@@ -40,6 +41,9 @@ export {
 	GetCredentialsForIdentityOutput,
 } from './AwsClients/CognitoIdentity';
 
+// Amplify-wide constructs
+export { UserProfile } from './types';
```
Comment (jimblanc) [inline packages/core/src/index.ts:45]: I think this should be exported from `libraryUtils.ts` to map to the `internals/utils` endpoint for now

File: packages/analytics/src/types/index.ts (hunk 1 above)
Comment (cshfang) [reply packages/analytics/src/types/index.ts:5]: It's probably preferable? But I was following established patterns from auth and storage...

File: packages/analytics/src/providers/pinpoint/apis/identifyUser.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+	PinpointIdentifyUserParameters,
+	UpdateEndpointException,
+} from '../types';
+import { resolveConfig, resolveCredentials } from '../utils';
+
+/**
+ * Identifies the current user with Pinpoint.
+ *
+ * @param {PinpointIdentifyUserParameters} params parameters used to construct requests sent to pinpoint service's
+ * UpdateEndpoint API.
```
Comment (jimblanc) [inline packages/analytics/src/providers/pinpoint/apis/identifyUser.ts:18]:
```suggestion
 * @param {PinpointIdentifyUserParameters} params parameters used to construct requests sent to Pinpoint's UpdateEndpoint API.
```

**Review by jimblanc — COMMENTED** [review]
Looks good to me, two nits

File: packages/analytics/__tests__/providers/pinpoint/utils/resolveCredentials.test.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+		identityId: 'identity-id',
+	};
+	// assert mocks
+	const mockFetchAuthSession = fetchAuthSession as jest.Mock;
+
+	beforeEach(() => {
+		mockFetchAuthSession.mockReset();
+	});
+
+	test('resolves required credentials', async () => {
```
Comment (ashwinkumar6) [inline packages/analytics/__tests__/providers/pinpoint/utils/resolveCredentials.test.ts:25]: NIT
QQ
Should we use `it` instead of `test`

Comment (cshfang) [reply packages/analytics/__tests__/providers/pinpoint/utils/resolveCredentials.test.ts:25]: 👍

File: packages/core/src/index.ts (hunk 2 above)
Comment (cshfang) [reply packages/core/src/index.ts:45]: Chatted offline and agreed to keep as is

**Review by jimblanc — APPROVED** [review]

File: packages/core/src/providers/pinpoint/types/pinpoint.ts [hunk 5]
```diff
@@ -18,19 +18,19 @@ export type PinpointProviderConfig = {
 	};
 };
 
-export type PinpointUserProfile = UserProfile & {
+export type PinpoinServiceOptions = {
```
Comment (jimblanc) [inline packages/core/src/providers/pinpoint/types/pinpoint.ts:21]:
```suggestion
export type PinpointServiceOptions = {
```

**Review by jimblanc — APPROVED** [review]
Makes sense to me

File: packages/core/src/providers/pinpoint/apis/updateEndpoint.ts [hunk 6]
```diff
@@ -14,11 +14,13 @@ import { cacheEndpointId, getEndpointId } from '../utils';
  * @internal
  */
 export const updateEndpoint = async ({
+	address,
```
Comment (Samaritan1011001) [inline packages/core/src/providers/pinpoint/apis/updateEndpoint.ts:17]: Can't comment on unchanged lines 🤯 
So if `address` and `optOut` isn't passed, is it fine for it to be undefined when making the service call? Just making sure.

File: packages/core/src/providers/pinpoint/types/pinpoint.ts [hunk 7]
```diff
@@ -18,19 +18,19 @@ export type PinpointProviderConfig = {
...
 	address?: string;
 	optOut?: 'ALL' | 'NONE';
 };
 
-export type PinpointUpdateEndpointParameters = {
+export type PinpointUpdateEndpointParameters = PinpoinServiceOptions & {
```
Comment (Samaritan1011001) [inline packages/core/src/providers/pinpoint/types/pinpoint.ts:26]: typo - `PinpoinServiceOptions`

**Review by Samaritan1011001 — APPROVED** [review]

File: packages/core/src/providers/pinpoint/apis/updateEndpoint.ts (hunk 6 above)
Comment (cshfang) [reply packages/core/src/providers/pinpoint/apis/updateEndpoint.ts:17]: Yeah it's a-ok

**Review by ashwinkumar6 — APPROVED** [review]
LGTM 
Thank you
